### DISCUSSION TITLE
Tooling for context sensitive Context-Menus generation?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-platform/eclipse.platform.ui#465] [Q&A]
I'd often like to add a menu item in a context sensitive way, e.g in this case "Show it only if project has a given nature and adapting it to X returns property y == some value".

This often becomes a nightmare (why the .... its not working?!?) and ends with a lots of try and error, poking around and copy snippets from here to there and in the end it is working only half-way because something was still missing.

I therefore wondering if there is some tooling that one can used that generates the required parts and is a bit more interactive than trial-and-error maybe even suggesting some usual filters, adapters and so on? 
And of course for debugging as well, I think even a printout like "your action was not enabled because condition X is false" would help a lot (especially as one easily can deadlock SWT when debugging menus...)

**Comment by mickaelistria** [comment]
Is it about context menu, or about evaluating/debugging Core Expressions in general?

**Reply by laeubi** [reply]
I think both, even though I found this most often useful for menu items (I think there is/was even a bug that core expressions do not work in all cases)...

**Comment by vogella** [comment]
E4 contributions can use imperitive expressions which allows you to use regular Java code and DI

**Reply by laeubi** [reply]
Yeah I use this for my RCP app and this is quite great... but never was able to get this working for "Eclipse", as there is always some "legacy" in the way that folls me around... but probably there is a tool where one can simply click on a menu item and gets E4 code copied to the clipboard that inserts a Menuitem right there?

For example I want a dynamic menu item where I rightclick on a resource then choose `New ...` and then lets say right after the `Project...` item with E4....

**Comment by vogella** [comment]
As a starter you can use the pde E4 menu  template which also adds a context menu

**Reply by laeubi** [reply]
@vogella it seems this only works when creating a new fragment, it would be good to have this fro existing fragments as well, anyways here again is too much "magic" where I think toolling support would be required, here is a screenshot to show the problem:
![grafik](https://user-images.githubusercontent.com/1331477/204725266-10f92fb0-b4bb-44d4-a7b0-886a3b71e0d3.png)

So what I'm looking for would be a button / tool I can activate (lets say instead of `Find` there is a button `Select`) where I can select an existing menu-item from the Running Eclipse and it fills out the magic strings accordingly. Because currently one can fill in anything, but wrong strings simply lead to "nothing" happen and I have not found an efficient way to even debug this.

**Reply by laeubi** [reply]
For example that is what [MenuSpy](https://www.vogella.com/tutorials/EclipseCodeAccess/article.html#menu-spy) gives me for `RightClick > New > <any item>`

![grafik](https://user-images.githubusercontent.com/1331477/208101963-288bd02b-bd6c-42bb-8c14-9fa8fef7db7d.png)

The "new" menu item is even not select-able with the spy because it is a sub menu...
